plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp

```diff
@@ -0,0 +1,172 @@
+
+      if(sri.end_block_num > 0) {
+         // if "end" is specified, it should be greater then start
+         EOS_ASSERT(sri.start_block_num <= sri.end_block_num, invalid_snapshot_request, "End block number should be greater or equal to start block number");
+         // if also block_spacing specified, check it
+         if(sri.block_spacing > 0) {
+            EOS_ASSERT(sri.start_block_num + sri.block_spacing <= sri.end_block_num, invalid_snapshot_request, "Block spacing exceeds defined by start and end range");
+         }
+      }
+
+      auto request_id = sri.snapshot_request_id ? sri.snapshot_request_id : _snapshot_id++;
+      _snapshot_requests.emplace(producer_plugin::snapshot_request_information{request_id, sri.block_spacing, sri.start_block_num, sri.end_block_num, {}});
+
+      auto& vec = _snapshot_requests.get<as_vector>();
+      std::vector<producer_plugin::snapshot_request_information> sr(vec.begin(), vec.end());
+      _snapshot_db << sr;
+   }
+
+   virtual void unschedule_snapshot(uint32_t sri) {
+      auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+      auto existing = snapshot_by_id.find(sri);
+      EOS_ASSERT(existing != snapshot_by_id.end(), snapshot_request_not_found, "Snapshot request not found");
+      _snapshot_requests.erase(existing);
+
+      auto& vec = _snapshot_requests.get<as_vector>();
+      std::vector<producer_plugin::snapshot_request_information> sr(vec.begin(), vec.end());
+      _snapshot_db << sr;
+   }
+
+   virtual get_snapshot_requests_result get_snapshot_requests() {
+      get_snapshot_requests_result result;
+      auto& asvector = _snapshot_requests.get<as_vector>();
+      result.requests.insert(result.requests.begin(), asvector.begin(), asvector.end());
+      return result;
+   }
+
+   // initialize with storage
+   void set_db_path(bfs::path db_path) {
+      _snapshot_db.set_path(db_path);
+      // init from db
+      if(fc::exists(_snapshot_db.get_json_path())) {
+         std::vector<producer_plugin::snapshot_request_information> sr;
+         _snapshot_db >> sr;
+         _snapshot_requests.insert(sr.begin(), sr.end());
+      }
+   }
+
+   // snapshot executor
+   void set_create_snapshot_fn(std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> fn) {
+      _create_snapshot = fn;
+   }
+
+
+   void execute_snapshot() {
+      auto next = [](const std::variant<fc::exception_ptr, producer_plugin::snapshot_information>& result) {
+         if(std::holds_alternative<fc::exception_ptr>(result)) {
+            try {
+               std::get<fc::exception_ptr>(result)->dynamic_rethrow_exception();
+            } catch(const fc::exception& e) {
+               edump((e.to_detail_string()));
+               throw;
+            } catch(const std::exception& e) {
+               edump((e.what()));
+               throw;
+            }
+         } else {
+            // success
+         }
+      };
+      _create_snapshot(next);
```

Comment:
Would like to see the `create_snapshot` code moved into this class and out of `producer_plugin`.